stdlib/3/subprocess.pyi

```diff
@@ -808,96 +808,347 @@ class CalledProcessError(Exception):
                  output: Optional[_TXT] = ...,
                  stderr: Optional[_TXT] = ...) -> None: ...
 
-class Popen:
+class _Popen(Generic[AnyStr]):
     args: _CMD
-    # We would like to give better types to these fields but currently
-    # have no way of overloading a constructor...
-    stdin: IO[Any]
-    stdout: IO[Any]
-    stderr: IO[Any]
+    stdin: IO[AnyStr]
+    stdout: IO[AnyStr]
+    stderr: IO[AnyStr]
     pid = 0
     returncode = 0
 
+    # Technically it is wrong that Popen provides __new__ instead of __init__
+    # but this shouldn't come up hopefully?
```

Comment:
Sure, the only way to know how important it is to be precise is to try it, so I propose to ahead.

Reply:
Yuck, so if anyone overrides `__init__` (like some of our code does to provide additional features), then the returned type is wrong.

What's the reason `__new__` needs to be used over `__init__?`

Reply:
I'm guessing the reason for using `__new__` is that with `__init__`, there wouldn't be a way to specify Popen's contained type because it doesn't come directly from any of the parameter types.

For what it's worth, the way pytype gets around this kind of problem is with a pyi extension we refer to as "mutations", e.g.:

```
class Popen:
  @overload
  def __init__(self, ...):
    self = Popen[str]
  @overload
  def __init__(self, ...):
    self = Popen[bytes]
```

On a related note, I just finished rolling out this change at Google, and while the `__new__`/`__init__` distinction caused a few false positives, the more precise Popen types flagged tons of places where subprocess output was being used in Python 3-incompatible ways, which was incredibly useful.